include/divergence_block_chain.h

```diff
@@ -0,0 +1,166 @@
+	size_t m_offset = 0;
+	size_t m_width = 0;
+};
+
+using task_hash = size_t;
+using task_hash_data = vector_2d<task_hash>;
+using divergence_map = std::unordered_map<task_hash, std::vector<node_id>>;
+
+/**
+ *  @brief This class checks for divergences of tasks between nodes.
+ *
+ *	It is responsible for collecting the task hashes from all nodes and checking for differences -> divergence.
+ *	When a divergence is found, the task record for the diverging task is printed and the program is terminated.
+ *  Additionally it also checks for deadlocks and prints a warning if one is detected.
+ */
+
+class divergence_block_chain {
+	friend struct divergence_block_chain_testspy;
+
+  public:
+	divergence_block_chain(const std::vector<task_record>& task_recorder, std::unique_ptr<communicator> comm)
+	    : m_local_nid(comm->get_local_nid()), m_num_nodes(comm->get_num_nodes()), m_sizes(comm->get_num_nodes()), m_task_recorder_window(task_recorder),
+	      m_communicator(std::move(comm)) {}
+
+	divergence_block_chain(const divergence_block_chain&) = delete;
+	divergence_block_chain(divergence_block_chain&&) = default;
+
+	~divergence_block_chain() = default;
+
+	divergence_block_chain& operator=(const divergence_block_chain&) = delete;
+	divergence_block_chain& operator=(divergence_block_chain&&) = delete;
+
+	bool check_for_divergence();
+
+  private:
+	node_id m_local_nid;
+	size_t m_num_nodes;
+
+	std::vector<task_hash> m_hashes;
+	std::vector<int> m_sizes;
+
+	sliding_window<task_record> m_task_recorder_window;
+
+	std::chrono::time_point<std::chrono::steady_clock> m_last_cleared = std::chrono::steady_clock::now();
+
+	std::unique_ptr<communicator> m_communicator;
+
+	void divergence_out(const divergence_map& check_map, const int task_num);
+
+	void add_new_hashes();
+	void clear(const int min_progress);
+	std::pair<int, int> collect_sizes();
+	task_hash_data collect_hashes(const int max_size);
+	divergence_map create_check_map(const task_hash_data& task_graphs, const int task_num) const;
+
+	void check_for_deadlock() const;
+
+	static void log_node_divergences(const divergence_map& check_map, const int task_num);
+
+	static void log_task_record(const divergence_map& check_map, const task_record& task, const task_hash hash);
+
+	void log_task_record_once(const divergence_map& check_map, const int task_num) const;
+};
+
+class divergence_sceduler {
+	friend struct runtime_testspy;
+
+  public:
+	divergence_sceduler(const std::vector<task_record>& task_recorder, std::unique_ptr<communicator> comm, bool test_mode = false)
+	    : m_block_chain(task_recorder, std::move(comm)) {
+		if(!test_mode) { start(); }
+	}
+
+	divergence_sceduler(const divergence_sceduler&) = delete;
+	divergence_sceduler(divergence_sceduler&&) = default;
+
+	divergence_sceduler& operator=(const divergence_sceduler&) = delete;
+	divergence_sceduler& operator=(divergence_sceduler&&) = delete;
+
+	~divergence_sceduler() { stop(); }
+
+  private:
+	void start() {
+		stop();
```

Comment:
??

Reply:
Again an oversight of my first implementation where it was possible to start the divergence checker two times, which had some different problems.